File: solutions/haskell/grains/1/src/Grains.hs
module Grains (square, total) where

square :: Integer -> Maybe Integer
square n = if (n > 64 || n <= 0) then Nothing else Just (2 ^ (n - 1))

total :: Integer
total = sum
  (map (\x -> case square x of 
                Nothing -> 0
                Just v -> v) [1..64])


      

File: solutions/c/grains/1/grains.c
#include "grains.h"
#include <math.h>



uint64_t square(uint8_t index)
{
        return pow(2, index - 1);
}

uint64_t total(void)
{    
    uint64_t all_squares = 0;    
    for (int i = 1; i < 65; i++)
    {
        all_squares += square(i);
    }
    return all_squares;
}